sunscreen_tfhe/src/gpu/tests/fft.rs

```diff
@@ -1,28 +1,134 @@
-use num::Complex;
+use bytemuck::Pod;
+use num::{Complex, Float, FromPrimitive, NumCast, Signed};
+use sunscreen_gpu_runtime::launch_kernel;
+
+use rustfft::FftPlanner;
 
 use crate::gpu::tests::get_runtimes;
 
-#[test]
-fn can_roundtrip_fft() {
+fn assert_complex_equalish<T: Float + NumCast + std::fmt::Display>(
+    actual: &Complex<T>,
+    expected: &Complex<T>,
+    eps: T,
+) {
+    let denom = if actual.re == T::from(0.0).unwrap() {
+        T::from(1.0).unwrap()
+    } else {
+        actual.re.abs()
+    };
```

Comment:
This seems to be alternating between a relative and absolute epsilon for checking for equality. It may be good to do two things:

1. Provide an explicit relative and absolute epsilon tolerance
2. The code may be easier in radial (magnitude and phase) coordinates. This should also be more robust to the meaning of the complex number since checking by components is checking for magnitude and phase directly; component wise the Cartesian error leads to different achieved tolerances for on the magnitude and phase based on the angle of the complex number.

Reply:
Alternatively you could copy what numpy does: https://github.com/numpy/numpy/blob/0532af47d6a815298b7841de00bdbc547104b237/numpy/testing/_private/utils.py#L503-L625

Reply:
Ideally, I'd like to express the error in number of ULPs, which is why I'm normalizing the error between 0 and 1. This should allow the caller to set a tolerance of `0.5 * op depth * f64::EPS`, but this isn't exactly a science yet. The GPU uses a sincos transcendental operation, which has 2 ULPs of error.